# Allo Anido A et al. (2025) Manuscript Analysis Pipelines

KDM5B-mediated transcriptional noise enables phenotypic plasticity and adaptive evolution in pediatric cancer

Phenotypic plasticity is a cancer hallmark, promoting heterogeneity and therapy evasion. However, little is known about the molecular mechanisms that drive this process. Transcriptional noise is the stochastic variability in gene expression and is a known source of phenotypic diversification in multiple organisms. We show that transcriptional noise, regulated by the histone demethylase KDM5, underlies intrinsic and adaptive phenotypic plasticity in pediatric cancers. Through functional and transcriptomic analyses, we demonstrate that transcriptional noise shapes both intrinsic state dynamics and plastic responses to therapy. Highly plastic bridge cell populations exhibit the highest levels of noise in neuroblastoma and hepatoblastoma. Following treatment, slowly proliferating drug-tolerant persister cells rely on noise-induced plasticity to awaken, increase the proliferation capacity of the cancer population, and reconstitute tumor heterogeneity. Suppression of KDM5B activity reduces noise and prevents awakening and subsequent recovery. These findings establish transcriptional noise as a key regulator of plasticity and suggest new avenues to target cancerâ€™s adaptive evolutionary strategies.

This Github repository details all scripts and code for processing, analysis and figures associated with Allo Anido et al., 2025 "KDM5B-mediated transcriptional noise enables phenotypic plasticity and adaptive evolution in pediatric cancer" 

For any questions or queries please contact - PreclinPaedCancerEvo@icr.ac.uk


# Analysis Pipelines:

The provided folders contain all code and scripts needed to process, analyse and plot all figures in the above mentioned manuscript including:

RNA Sequencing (MemorySeq) \
Transcriptional noise quantification (BASiCS) 
